ppoprf/src/ppoprf.rs

```diff
@@ -118,6 +118,14 @@ impl ServerPublicKey {
     let md = md_pk.decompress().unwrap();
     Ok(Point::from(b + md))
   }
+
+  pub fn serialize_to_bincode(&self) -> Result<Vec<u8>, bincode::Error> {
+    bincode::serialize(self)
+  }
+
+  pub fn load_from_bincode(data: &[u8]) -> Result<Self, bincode::Error> {
+    bincode::deserialize(data)
```

Comment:
Might as well set a limit here. 1+ 32 + 1 + 256*(1 + 32) should be correct?

Reply:
Fortunately, bincode seems to read only what is needed! 

https://github.com/bincode-org/bincode/blob/6ec69a38b5698876662e93ca928f6f41183611d5/src/lib.rs#L143-L147

Reply:
My concern was
```rust
{
    let buffer = "overflow".as_bytes();
    let bad = unsafe {
        std::slice::from_raw_parts(buffer.as_ptr(), usize::MAX)
    };
    // panic!
    let _test: Vec<u8> = bincode::deserialize(bad).expect("bincode failure");
}
```

Reply:
In your example, I would imagine the program would crash after the unsafe code is executed

Reply:
Yes exactly. I can also make it panic with allocation failure (decoding into a string, which seems to reserve the full slice length) or read some smaller amount past the end of the real data and stuff in into the key HashMap where an attacker might be able to exfiltrate it.

It's generally good practice to guard arbitrarily-large reads where possible. It also might have caught base64 issue, since the maximum expected length wouldn't be enough for the actual encoding if you tested with 256 keys.

Reply:
To clarify, I mean that it could crash before the deserialize method is called. Could you clarify what you mean by the base64 issue? Do you mean handling cases where the input byte data length is too small to decode?

Bincode handles such cases already. When populating collections such as `Vec` or `HashMap`, the data is internally prefixed with the number of entries for that particular collection. If the input data is unexpectedly truncated, then bincode will return an error. Bincode also handles byte data with lengths that are too large, by reading only what is necessary to create the struct.

Reply:
> To clarify, I mean that it could crash before the deserialize method is called.

Where do you see it crashing? `gdb` shows that fragment segfaulting inside `bincode::deserialize()`. That's the point where unallocated memory is accessed. But my real point is that imposing a read bound helps limit the damage attacker-controlled input can cause.

> Could you clarify what you mean by the base64 issue?

I just mean that your patch is base64-encoding the `Point` values before bincoding them. Unless there's some reason you intended that behaviour?

Reply:
I wasn't sure how `from_raw_parts` worked under the hood, so I assumed it would perform some data access when called. After testing, that doesn't seem to be the case and it fails at the bincode step.

It's hard to guard against the many different cases of unsafe code, but I added the check nonetheless. The limit has a bit of padding to handle future changes.

Completely forgot that we had base64 encoding for the ristretto in the Point. I moved it up to the Evaluation struct so the serialization only applies to the example server response. Please review this when you have a moment.

Reply:
Yes, it's using a fixed-length encoding for the points as 32 bytes now.